Error in match statement generated when no compile-time condition

Hello! I use this crate because I am very comfortable with `Bound parameters`!

I have been using it recently and share it with you because I had an error when updating from `v0.1.4` to `v0.2.1`.

## Overview

If `conditional_query_as` is used in a form where the condition after the third argument is not written and only `Compile-time bound parameters` are used, the pattern match statement for the condition generated by the macro is `mismatched type`.

## Details

Rust Version: 1.83.0
OS: macos

### Codes

The actual code is here.

<details>
<summary>code</summary>

```rust
use sqlx::types::chrono::NaiveDateTime;
use sqlx::Connection;
use sqlx::postgres::PgConnection;
use sqlx_conditional_queries::conditional_query_as;

#[derive(Debug)]
struct Count {
    count: i64,
}

#[tokio::main]
async fn main() -> Result<(), sqlx::Error> {
    let connection_string = "postgresql://postgres:<password>@localhost:10000/postgres";

    let mut connection = PgConnection::connect(connection_string).await?;

    let created_at = NaiveDateTime::parse_from_str("2025-01-06 00:00:00", "%Y-%m-%d %H:%M:%S").unwrap();

    let result: Count = conditional_query_as!(
        Count,
        r#"
            SELECT COUNT(*) AS "count!"
            FROM contents
            WHERE created_at > {created_at}
        "#
    )
    .fetch_one(&mut connection)
    .await?;

    println!("{result:?}");

    Ok(())
}
```

The error message will appear as follows.

```text
error[E0308]: mismatched types
  --> mini-rust/example/src/main.rs:19:25
   |
19 |       let result: Count = conditional_query_as!(
   |  _________________________^
20 | |         Count,
21 | |         r#"
22 | |             SELECT COUNT(*) AS "count!"
...  |
25 | |         "#
26 | |     )
   | |     ^
   | |     |
   | |_____expected `((),)`, found `()`
   |       this expression has type `((),)`
   |
   = note:  expected tuple `((),)`
           found unit type `()`
   = note: this error originates in the macro `conditional_query_as` (in Nightly builds, run with -Z macro-backtrace for more info)
```

By the way, the table is prepared as follows.

```sql
                                Table "public.contents"
   Column   |              Type              | Collation | Nullable |      Default      
------------+--------------------------------+-----------+----------+-------------------
 id         | uuid                           |           | not null | 
 name       | character varying(512)         |           | not null | 
 created_at | timestamp(3) without time zone |           | not null | CURRENT_TIMESTAMP
 updated_at | timestamp(3) without time zone |           | not null | CURRENT_TIMESTAMP
Indexes:
    "contents_pkey" PRIMARY KEY, btree (id)
```

`Cargo.toml` is defined as follows.

```toml
[package]
name = "example"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1.42.0", features = [ "full" ] }
sqlx = { version = "0.8.3", default-features = false, features = [ "chrono", "macros", "postgres", "runtime-tokio" ] }
sqlx-conditional-queries = { version = "0.2.1", features = ["postgres"] }
```


</details>


### Expand

It appeared that a mismatched type was occurring in the code expanded by the macro, so I expanded it using `cargo expand`.

<details>
<summary>`v0.1.4`</summary>

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use sqlx::types::chrono::NaiveDateTime;
use sqlx::Connection;
use sqlx::postgres::PgConnection;
use sqlx_conditional_queries::conditional_query_as;
struct Count {
    count: i64,
}
#[automatically_derived]
impl ::core::fmt::Debug for Count {
    #[inline]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        ::core::fmt::Formatter::debug_struct_field1_finish(
            f,
            "Count",
            "count",
            &&self.count,
        )
    }
}
fn main() -> Result<(), sqlx::Error> {
    let body = async {
        let connection_string = "postgresql://postgres:postgres@localhost:10000/postgres";
        let mut connection = PgConnection::connect(connection_string).await?;
        let created_at = NaiveDateTime::parse_from_str(
                "2025-01-06 00:00:00",
                "%Y-%m-%d %H:%M:%S",
            )
            .unwrap();
        let result: Count = {
            enum ConditionalMap<'q, DB: ::sqlx::Database, A, F0> {
                Variant0(::sqlx::query::Map<'q, DB, F0, A>),
            }
            impl<'q, DB, A, O, F0> ConditionalMap<'q, DB, A, F0>
            where
                DB: ::sqlx::Database,
                A: 'q + ::sqlx::IntoArguments<'q, DB> + ::std::marker::Send,
                O: ::std::marker::Unpin + ::std::marker::Send,
                F0: ::std::ops::FnMut(DB::Row) -> ::sqlx::Result<O>
                    + ::std::marker::Send,
            {
                /// See [`sqlx::query::Map::fetch`]
                pub fn fetch<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx_conditional_queries::exports::BoxStream<
                    'e,
                    ::sqlx::Result<O>,
                >
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch(executor),
                    }
                }
                /// See [`sqlx::query::Map::fetch_many`]
                #[deprecated = "Only the SQLite driver supports multiple statements in one prepared statement and that behavior is deprecated. Use `sqlx::raw_sql()` instead. See https://github.com/launchbadge/sqlx/issues/3108 for discussion."]
                pub fn fetch_many<'e, 'c: 'e, E>(
                    mut self,
                    executor: E,
                ) -> ::sqlx_conditional_queries::exports::BoxStream<
                    'e,
                    ::sqlx::Result<::sqlx::Either<DB::QueryResult, O>>,
                >
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                {
                    match self {
                        Self::Variant0(map) => {
                            #[allow(deprecated)]
                            map.fetch_many(executor)
                        }
                    }
                }
                /// See [`sqlx::query::Map::fetch_all`]
                pub async fn fetch_all<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<::std::vec::Vec<O>>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_all(executor).await,
                    }
                }
                /// See [`sqlx::query::Map::fetch_one`]
                pub async fn fetch_one<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<O>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_one(executor).await,
                    }
                }
                /// See [`sqlx::query::Map::fetch_optional`]
                pub async fn fetch_optional<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<::std::option::Option<O>>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_optional(executor).await,
                    }
                }
            }
            match ((),) {
                ((),) => {
                    ConditionalMap::Variant0({
                        {
                            #[allow(clippy::all)]
                            {
                                use ::sqlx::Arguments as _;
                                let arg0 = &(created_at);
                                #[allow(clippy::missing_panics_doc, clippy::unreachable)]
                                if false {
                                    use ::sqlx::ty_match::{
                                        WrapSameExt as _, MatchBorrowExt as _,
                                    };
                                    let expr = ::sqlx::ty_match::dupe_value(arg0);
                                    let ty_check = ::sqlx::ty_match::WrapSame::<
                                        sqlx::types::chrono::NaiveDateTime,
                                        _,
                                    >::new(&expr)
                                        .wrap_same();
                                    let (mut _ty_check, match_borrow) = ::sqlx::ty_match::MatchBorrow::new(
                                        ty_check,
                                        &expr,
                                    );
                                    _ty_check = match_borrow.match_borrow();
                                    ::core::panicking::panic(
                                        "internal error: entered unreachable code",
                                    );
                                }
                                let mut query_args = <sqlx::postgres::Postgres as ::sqlx::database::Database>::Arguments::<
                                    '_,
                                >::default();
                                query_args
                                    .reserve(
                                        1usize,
                                        0
                                            + ::sqlx::encode::Encode::<
                                                sqlx::postgres::Postgres,
                                            >::size_hint(arg0),
                                    );
                                let query_args = ::core::result::Result::<
                                    _,
                                    ::sqlx::error::BoxDynError,
                                >::Ok(query_args)
                                    .and_then(move |mut query_args| {
                                        query_args.add(arg0).map(move |()| query_args)
                                    });
                                ::sqlx::__query_with_result::<
                                    sqlx::postgres::Postgres,
                                    _,
                                >(
                                        "\n            SELECT COUNT(*) AS \"count!\"\n            FROM contents\n            WHERE created_at > $1\n        ",
                                        query_args,
                                    )
                                    .try_map(|row: sqlx::postgres::PgRow| {
                                        use ::sqlx::Row as _;
                                        #[allow(non_snake_case)]
                                        let sqlx_query_as_count = row
                                            .try_get_unchecked::<i64, _>(0usize)?
                                            .into();
                                        ::std::result::Result::Ok(Count {
                                            count: sqlx_query_as_count,
                                        })
                                    })
                            }
                        }
                    })
                }
            }
        }
            .fetch_one(&mut connection)
            .await?;
        {
            ::std::io::_print(format_args!("{0:?}\n", result));
        };
        Ok(())
    };
    #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
    {
        return tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()
            .expect("Failed building the Runtime")
            .block_on(body);
    }
}
```

</details>



<details>
<summary>`v0.2.1`</summary>

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use sqlx::types::chrono::NaiveDateTime;
use sqlx::Connection;
use sqlx::postgres::PgConnection;
use sqlx_conditional_queries::conditional_query_as;
struct Count {
    count: i64,
}
#[automatically_derived]
impl ::core::fmt::Debug for Count {
    #[inline]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        ::core::fmt::Formatter::debug_struct_field1_finish(
            f,
            "Count",
            "count",
            &&self.count,
        )
    }
}
fn main() -> Result<(), sqlx::Error> {
    let body = async {
        let connection_string = "postgresql://postgres:postgres@localhost:10000/postgres";
        let mut connection = PgConnection::connect(connection_string).await?;
        let created_at = NaiveDateTime::parse_from_str(
                "2025-01-06 00:00:00",
                "%Y-%m-%d %H:%M:%S",
            )
            .unwrap();
        let result: Count = {
            enum ConditionalMap<'q, DB: ::sqlx::Database, A, F0, F1> {
                Variant0(::sqlx::query::Map<'q, DB, F0, A>),
                Variant1(::sqlx::query::Map<'q, DB, F1, A>),
            }
            impl<'q, DB, A, O, F0, F1> ConditionalMap<'q, DB, A, F0, F1>
            where
                DB: ::sqlx::Database,
                A: 'q + ::sqlx::IntoArguments<'q, DB> + ::std::marker::Send,
                O: ::std::marker::Unpin + ::std::marker::Send,
                F0: ::std::ops::FnMut(DB::Row) -> ::sqlx::Result<O>
                    + ::std::marker::Send,
                F1: ::std::ops::FnMut(DB::Row) -> ::sqlx::Result<O>
                    + ::std::marker::Send,
            {
                /// See [`sqlx::query::Map::fetch`]
                pub fn fetch<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx_conditional_queries::exports::BoxStream<
                    'e,
                    ::sqlx::Result<O>,
                >
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                    F1: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch(executor),
                        Self::Variant1(map) => map.fetch(executor),
                    }
                }
                /// See [`sqlx::query::Map::fetch_many`]
                #[deprecated = "Only the SQLite driver supports multiple statements in one prepared statement and that behavior is deprecated. Use `sqlx::raw_sql()` instead. See https://github.com/launchbadge/sqlx/issues/3108 for discussion."]
                pub fn fetch_many<'e, 'c: 'e, E>(
                    mut self,
                    executor: E,
                ) -> ::sqlx_conditional_queries::exports::BoxStream<
                    'e,
                    ::sqlx::Result<::sqlx::Either<DB::QueryResult, O>>,
                >
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                    F1: 'e,
                {
                    match self {
                        Self::Variant0(map) => {
                            #[allow(deprecated)]
                            map.fetch_many(executor)
                        }
                        Self::Variant1(map) => {
                            #[allow(deprecated)]
                            map.fetch_many(executor)
                        }
                    }
                }
                /// See [`sqlx::query::Map::fetch_all`]
                pub async fn fetch_all<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<::std::vec::Vec<O>>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                    F1: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_all(executor).await,
                        Self::Variant1(map) => map.fetch_all(executor).await,
                    }
                }
                /// See [`sqlx::query::Map::fetch_one`]
                pub async fn fetch_one<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<O>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                    F1: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_one(executor).await,
                        Self::Variant1(map) => map.fetch_one(executor).await,
                    }
                }
                /// See [`sqlx::query::Map::fetch_optional`]
                pub async fn fetch_optional<'e, 'c: 'e, E>(
                    self,
                    executor: E,
                ) -> ::sqlx::Result<::std::option::Option<O>>
                where
                    'q: 'e,
                    E: 'e + ::sqlx::Executor<'c, Database = DB>,
                    DB: 'e,
                    O: 'e,
                    F0: 'e,
                    F1: 'e,
                {
                    match self {
                        Self::Variant0(map) => map.fetch_optional(executor).await,
                        Self::Variant1(map) => map.fetch_optional(executor).await,
                    }
                }
            }
            match ((),) {
                () => {
                    ConditionalMap::Variant0({
                        {
                            #[allow(clippy::all)]
                            {
                                use ::sqlx::Arguments as _;
                                let arg0 = &(created_at);
                                #[allow(clippy::missing_panics_doc, clippy::unreachable)]
                                if false {
                                    use ::sqlx::ty_match::{
                                        WrapSameExt as _, MatchBorrowExt as _,
                                    };
                                    let expr = ::sqlx::ty_match::dupe_value(arg0);
                                    let ty_check = ::sqlx::ty_match::WrapSame::<
                                        sqlx::types::chrono::NaiveDateTime,
                                        _,
                                    >::new(&expr)
                                        .wrap_same();
                                    let (mut _ty_check, match_borrow) = ::sqlx::ty_match::MatchBorrow::new(
                                        ty_check,
                                        &expr,
                                    );
                                    _ty_check = match_borrow.match_borrow();
                                    ::core::panicking::panic(
                                        "internal error: entered unreachable code",
                                    );
                                }
                                let mut query_args = <sqlx::postgres::Postgres as ::sqlx::database::Database>::Arguments::<
                                    '_,
                                >::default();
                                query_args
                                    .reserve(
                                        1usize,
                                        0
                                            + ::sqlx::encode::Encode::<
                                                sqlx::postgres::Postgres,
                                            >::size_hint(arg0),
                                    );
                                let query_args = ::core::result::Result::<
                                    _,
                                    ::sqlx::error::BoxDynError,
                                >::Ok(query_args)
                                    .and_then(move |mut query_args| {
                                        query_args.add(arg0).map(move |()| query_args)
                                    });
                                ::sqlx::__query_with_result::<
                                    sqlx::postgres::Postgres,
                                    _,
                                >(
                                        "\n            SELECT COUNT(*) AS \"count!\"\n            FROM contents\n            WHERE created_at > $1\n        ",
                                        query_args,
                                    )
                                    .try_map(|row: sqlx::postgres::PgRow| {
                                        use ::sqlx::Row as _;
                                        #[allow(non_snake_case)]
                                        let sqlx_query_as_count = row
                                            .try_get_unchecked::<i64, _>(0usize)?
                                            .into();
                                        ::std::result::Result::Ok(Count {
                                            count: sqlx_query_as_count,
                                        })
                                    })
                            }
                        }
                    })
                }
                ((),) => {
                    ConditionalMap::Variant1({
                        {
                            #[allow(clippy::all)]
                            {
                                use ::sqlx::Arguments as _;
                                let arg0 = &(created_at);
                                #[allow(clippy::missing_panics_doc, clippy::unreachable)]
                                if false {
                                    use ::sqlx::ty_match::{
                                        WrapSameExt as _, MatchBorrowExt as _,
                                    };
                                    let expr = ::sqlx::ty_match::dupe_value(arg0);
                                    let ty_check = ::sqlx::ty_match::WrapSame::<
                                        sqlx::types::chrono::NaiveDateTime,
                                        _,
                                    >::new(&expr)
                                        .wrap_same();
                                    let (mut _ty_check, match_borrow) = ::sqlx::ty_match::MatchBorrow::new(
                                        ty_check,
                                        &expr,
                                    );
                                    _ty_check = match_borrow.match_borrow();
                                    ::core::panicking::panic(
                                        "internal error: entered unreachable code",
                                    );
                                }
                                let mut query_args = <sqlx::postgres::Postgres as ::sqlx::database::Database>::Arguments::<
                                    '_,
                                >::default();
                                query_args
                                    .reserve(
                                        1usize,
                                        0
                                            + ::sqlx::encode::Encode::<
                                                sqlx::postgres::Postgres,
                                            >::size_hint(arg0),
                                    );
                                let query_args = ::core::result::Result::<
                                    _,
                                    ::sqlx::error::BoxDynError,
                                >::Ok(query_args)
                                    .and_then(move |mut query_args| {
                                        query_args.add(arg0).map(move |()| query_args)
                                    });
                                ::sqlx::__query_with_result::<
                                    sqlx::postgres::Postgres,
                                    _,
                                >(
                                        "\n            SELECT COUNT(*) AS \"count!\"\n            FROM contents\n            WHERE created_at > $1\n        ",
                                        query_args,
                                    )
                                    .try_map(|row: sqlx::postgres::PgRow| {
                                        use ::sqlx::Row as _;
                                        #[allow(non_snake_case)]
                                        let sqlx_query_as_count = row
                                            .try_get_unchecked::<i64, _>(0usize)?
                                            .into();
                                        ::std::result::Result::Ok(Count {
                                            count: sqlx_query_as_count,
                                        })
                                    })
                            }
                        }
                    })
                }
            }
        }
            .fetch_one(&mut connection)
            .await?;
        {
            ::std::io::_print(format_args!("{0:?}\n", result));
        };
        Ok(())
    };
    #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
    {
        return tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()
            .expect("Failed building the Runtime")
            .block_on(body);
    }
}
```

</details>

Comparison shows that the number of values generated by the `ConditionalMap` of the enum is different.
In `v0.1.4` there is one and in `v0.2.1` there are two.
In other words, it behaves as if there is some condition, even though there is no condition.

```rust
// v0.1.4
enum ConditionalMap<'q, DB: ::sqlx::Database, A, F0> {
    Variant0(::sqlx::query::Map<'q, DB, F0, A>),
}

// v0.2.1
enum ConditionalMap<'q, DB: ::sqlx::Database, A, F0, F1> {
  Variant0(::sqlx::query::Map<'q, DB, F0, A>),
  Variant1(::sqlx::query::Map<'q, DB, F1, A>),
}
```

And in the subsequent branching of query statement generation by match, for `v.0.2.1`, it seemed to be trying to branch on `()` for the condition `((),)`.

```rust
// v.0.1.4
            match ((),) {
                ((),) => {
                    ConditionalMap::Variant0({
          ...

// v.0.2.1
            match ((),) {
                () => { // Probably a typing error here. 
                    ConditionalMap::Variant0({
                    ....
                ((),) => {
                    ConditionalMap::Variant0({
          ...
```


## Summary

Although I feel that the use of no conditions is an extended use case, I consider `Run-time bound parameters` to be very beneficial in terms of readability.

Thank you for creating a very useful library! I hope this Issue will help you to improve it!

--- a/core/src/lower.rs
+++ b/core/src/lower.rs
@@ -83,6 +83,11 @@ pub(crate) fn lower(analyzed: AnalyzedConditionalQueryAs) -> LoweredConditionalQ
     // keep track of which match expressions belongs to which part of the match arm's expression.
     let mut match_arms = Vec::new();
     for binding in bindings.into_iter().multi_cartesian_product() {
+        // `multi_cartesian_product` returns one empty `Vec` if the iterator was empty.
+        if binding.is_empty() {
+            continue;
+        }
+
         let mut guards = Vec::new();
         let mut bindings = HashMap::new();
         binding.into_iter().for_each(|(g, b)| {

